src/H5FDmulti.c

```diff
@@ -2068,6 +2071,64 @@ H5FD_multi_delete(const char *filename, hid_t fapl_id)
 } /* end H5FD_multi_delete() */
 H5_MULTI_GCC_DIAG_ON("format-nonliteral")
 
+/*-------------------------------------------------------------------------
+ * Function:    H5FD_multi_ctl
+ *
+ * Purpose:     Multi VFD version of the ctl callback.
+ *
+ *              The desired operation is specified by the op_code
+ *              parameter.
+ *
+ *              The flags parameter controls management of op_codes that
+ *              are unknown to the callback
+ *
+ *              The input and output parameters allow op_code specific
+ *              input and output
+ *
+ *              At present, this VFD supports no op codes of its own and
+ *              simply passes ctl calls on to each member VFD.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5FD_multi_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void *input, void **output)
```

Comment:
@jrmainzer The terminology for what constitutes a terminal vs passthrough VFD seemed to get a little bit fuzzy for certain VFDs. At first, it seemed like the splitter VFD was really the only passthrough VFD that we currently have. However, the multi VFD also seems like it might fit the bill, so I went ahead and implemented the ctl function here as though the multi VFD were a passthrough VFD. If that's not the case, I'm happy to change it back/remove the ctl callback from the multi VFD.

Reply:
Actually, using the definition of terminal vs. passthrough VFD in H5FDpublic.h:

Terminal VFD: Lowest VFD in the VFD stack through which all VFD calls pass.  Note that this definition is situational.  For example, the sec2 VFD is typically terminal.  However, in the context of the family file driver, it is not -- the family file driver is the bottom VFD through which all VFD calls pass, and thus it is terminal.

Pass through VFD:  Any VFD that relays all VFD calls (with the possible exception of some non-I/O related calls) to underlying VFD(s).

The multi-file driver is terminal, not pass through.  In contrast splitter is a pass through VFDs.  When and if we implement a proper network VFD (i.e. mirror with R/W instead of just W/O) it will be a pass through VFD as well.

None of this is to say that the above definitions are what we should go with in the long term.  They are very much a first cut, and we should change them if they become inconvenient.

Reply:
This is interesting; is there an example of what API calls the multi driver doesn't pass down? It seems as though it passes all VFD calls down to all of its underlying member VFDs.

Reply:
I guess I didn't express this well.  Yes, the multi-file driver will pass all VFD calls down, but none of its underlying VFDs will receive all calls.  Thus the multi-file driver is terminal because all calls pass through it, and no underlying VFD receives all calls.  Does this make sense?

Reply:
Hmm, based just on the definition from H5FDpublic.h, then yes I can see how we could consider the splitter VFD as being the only passthrough VFD. Intuitively though, it seems to me like VFDs which use the public versions of certain H5FD calls (H5FDopen/H5FDread/H5FDwrite, etc.) could be considered as passthrough VFDs. However, if our current definitions do not consider the multi VFD to be a passthrough VFD, then I should either modify its ctl callback to not handle the "route to terminal VFD" flag, or remove the callback altogether.

Reply:
That or change the definition to something more useful.  That said, my take would be to keep the multi file driver ctl function as a stub for later use.  

Note that we will probably develop a more nuanced selection of routing flags.  For example, if we were to use the split file driver, with the onion VFD on the metadata channel, we might want to be able to route a ctl call to the metadata channel to tell the onion VFD to start a new version -- thus giving us a way of rolling back to the last consistent view of metadata in the event of a crash.  I'm sure you can think of other variations.

The bottom line is that this ctl call and its associated flags and op-codes will likely evolve over time.  if so, we will probably find ourselves changing and/or extending our definitions of pass through vs. terminal VFDs.

Reply:
> On Wed, Sep 01, 2021 at 12:04:00AM +0000, jrmainzer wrote: @jrmainzer commented on this pull request. > + * + * The flags parameter controls management of op_codes that + * are unknown to the callback + * + * The input and output parameters allow op_code specific + * input and output + * + * At present, this VFD supports no op codes of its own and + * simply passes ctl calls on to each member VFD. + * + * Return: Non-negative on success/Negative on failure + * + *------------------------------------------------------------------------- + */ +static herr_t +H5FD_multi_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void *input, void **output) I guess I didn't express this well. Yes, the multi-file driver will pass all VFD calls down, but none of its underlying VFDs will receive all calls. Thus the multi-file driver is terminal because all calls pass through it, and no underlying VFD receives all calls. Does this make sense?
> No, this does not make sense. It defies any simple definition of terminal. Dave

I agree with Dave. That definition is really confusing. I view the multi VFD as pass-through. If anything, we could have pass-through VFDs that solely forward and "demultiplexing" VFDs that route VFD calls to an array of underlying VFDs.

Reply:
Please suggest a more useful set of terminology to classify VFDs.